File: src/draw.cpp
/* ahi is a simple snake game written in C++
 *
 * Copyright (C) 2025  Farid Farajli
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <cassert>
#include <format>
#include <ranges>

#include "sdl.h"

#include "draw.h"

struct Color {
	std::uint8_t r{};
	std::uint8_t g{};
	std::uint8_t b{};
	std::uint8_t a{};
	constexpr Color(const char *s)
	{
		constexpr auto hex = [](char ch)
		{
			return ch > 'A' ? ch-'A'+10 : ch-'0';
		};

		assert(s[0] == '#');
		assert(s[1] >= '0' && s[1] <= 'F');
		assert(s[2] >= '0' && s[2] <= 'F');
		assert(s[3] >= '0' && s[3] <= 'F');
		assert(s[4] >= '0' && s[4] <= 'F');
		assert(s[5] >= '0' && s[5] <= 'F');
		assert(s[6] >= '0' && s[6] <= 'F');
		assert(s[7] >= '0' && s[7] <= 'F');
		assert(s[8] >= '0' && s[8] <= 'F');
		assert(!s[9]);

		this->r = hex(s[1])*16 + hex(s[2]);
		this->g = hex(s[3])*16 + hex(s[4]);
		this->b = hex(s[5])*16 + hex(s[6]);
		this->a = hex(s[7])*16 + hex(s[8]);
	}
	constexpr SDL_Color sdl_color(this const Color &self)
	{
		return SDL_Color{
			.r = self.r,
			.g = self.g,
			.b = self.b,
			.a = self.a,
		};
	}
};

constexpr Color ColorBackground = "#BEBEBEFF";
constexpr Color ColorBody = "#FFFF00FF";
constexpr Color ColorHead = "#00FF00FF";
constexpr Color ColorGrid = "#000000FF";
constexpr Color ColorFood = "#0000FFFF";
constexpr Color ColorScore = "#000000FF";

bool set_color(Color c)
{
	return SDL_SetRenderDrawColor(renderer, c.r, c.g, c.b, c.a);
}

bool draw_cell(Position pos)
{
	SDL_FRect rect{
		.x = float(pos.x*(Factor+LineWidth)),
		.y = float(pos.y*(Factor+LineWidth)),
		.w = Factor,
		.h = Factor,
	};

	return SDL_RenderFillRect(renderer, &rect);
}

void draw_food(Position food)
{
	assert(set_color(ColorFood));
	assert(draw_cell(food));
}

void draw_grid()
{
	SDL_FRect rect{};

	assert(set_color(ColorGrid));

	for (auto line : vs::iota(std::uint64_t{}, ColumnCount-std::uint64_t{1})) {
		rect.x = line*(Factor+LineWidth)+Factor;
		rect.y = 0.0f;
		rect.w = LineWidth;
		rect.h = GridHeight;
		assert(SDL_RenderFillRect(renderer, &rect));
	}

	for (auto line : vs::iota(std::uint64_t{}, RowCount)) {
		rect.x = 0.0f;
		rect.y = line*(Factor+LineWidth)+Factor;
		rect.w = GridWidth;
		rect.h = LineWidth;
		assert(SDL_RenderFillRect(renderer, &rect));
	}
}

void draw_snake(const Snake &snake)
{
	auto head {snake.body.front()};

	assert(set_color(ColorBody));
	for (auto segment : snake.body | vs::take(snake.length) | vs::drop(1))
		assert(draw_cell(segment));

	assert(set_color(ColorHead));
	assert(draw_cell(head));
}

void draw_score(std::uint64_t score)
{
	SDL_FRect rect{
		.x = 0.0f,
		.y = (Factor+LineWidth)*RowCount,
		.w = ScoreWidth,
		.h = ScoreHeight,
	};
	SDL_Surface *surface{};
	SDL_Texture *texture{};

	surface = TTF_RenderText_Solid(font, std::format("{:03}", score).c_str(), 0, ColorScore.sdl_color());
	assert(surface);
	texture = SDL_CreateTextureFromSurface(renderer, surface);
	assert(texture);
	assert(SDL_RenderTexture(renderer, texture, NULL, &rect));
	SDL_DestroyTexture(texture);
	SDL_DestroySurface(surface);
}

void draw(const World &w)
{
	assert(set_color(ColorBackground));
	assert(SDL_RenderClear(renderer));

	draw_grid();

	draw_snake(w.snake);

	draw_food(w.food);

	draw_score(w.score);

	assert(SDL_RenderPresent(renderer));
}

File: src/draw.rs
use sdl3::pixels::Color;
use sdl3::rect::Rect;
use sdl3::render::{Canvas, RenderTarget};

use crate::world::{COLUMN_COUNT, Position, ROW_COUNT, Snake, World};

const CELL_WIDTH: u16 = 90;
const LINE_WIDTH: u16 = 1;
pub const GRID_WIDTH: u16 = COLUMN_COUNT * CELL_WIDTH + (COLUMN_COUNT - 1) * LINE_WIDTH;
pub const GRID_HEIGHT: u16 = ROW_COUNT * CELL_WIDTH + ROW_COUNT * LINE_WIDTH;

const COLOR_BACKGROUND: Color = Color::RGBA(0xBE, 0xBE, 0xBE, 0xFF);
const COLOR_GRID: Color = Color::RGBA(0x00, 0x00, 0x00, 0xFF);
const COLOR_HEAD: Color = Color::RGBA(0x00, 0xFF, 0x00, 0xFF);
const COLOR_BODY: Color = Color::RGBA(0xFF, 0xFF, 0x00, 0xFF);
const COLOR_FOOD: Color = Color::RGBA(0x00, 0x00, 0xFF, 0xFF);

fn draw_grid<T: RenderTarget>(canvas: &mut Canvas<T>) {
    let mut rect = Rect::new(0, 0, 0, 0);

    canvas.set_draw_color(COLOR_GRID);

    for line in 0..COLUMN_COUNT - 1 {
        rect.set_x(i32::from(line * (CELL_WIDTH + LINE_WIDTH) + CELL_WIDTH));
        rect.set_y(0);
        rect.set_width(LINE_WIDTH.into());
        rect.set_height(GRID_HEIGHT.into());
        canvas.fill_rect(rect).unwrap();
    }

    for line in 0..ROW_COUNT {
        rect.set_x(0);
        rect.set_y(i32::from(line * (CELL_WIDTH + LINE_WIDTH) + CELL_WIDTH));
        rect.set_width(GRID_WIDTH.into());
        rect.set_height(LINE_WIDTH.into());
        canvas.fill_rect(rect).unwrap();
    }
}

fn draw_cell<T: RenderTarget>(canvas: &mut Canvas<T>, pos: Position) {
    let rect = Rect::new(
        i32::from(pos.x * (CELL_WIDTH + LINE_WIDTH)),
        i32::from(pos.y * (CELL_WIDTH + LINE_WIDTH)),
        CELL_WIDTH.into(),
        CELL_WIDTH.into(),
    );

    canvas.fill_rect(rect).unwrap();
}

fn draw_snake<T: RenderTarget>(canvas: &mut Canvas<T>, snake: &Snake) {
    canvas.set_draw_color(COLOR_HEAD);
    draw_cell(canvas, snake.body[0]);

    canvas.set_draw_color(COLOR_BODY);
    for segment in &snake.body[1..] {
        draw_cell(canvas, *segment);
    }
}

fn draw_food<T: RenderTarget>(canvas: &mut Canvas<T>, food: Position) {
    canvas.set_draw_color(COLOR_FOOD);
    draw_cell(canvas, food);
}

pub fn draw_scene<T: RenderTarget>(canvas: &mut Canvas<T>, w: &World) {
    canvas.set_draw_color(COLOR_BACKGROUND);
    canvas.clear();

    draw_grid(canvas);
    draw_snake(canvas, &w.snake);
    draw_food(canvas, w.food);

    canvas.present();
}
